src/autobattle.cpp

```diff
@@ -0,0 +1,376 @@
+#include "game_battlealgorithm.h"
+#include "game_battle.h"
+#include "algo.h"
+#include "player.h"
+#include "output.h"
+#include "rand.h"
+#include <lcf/reader_util.h>
+#include <lcf/data.h>
+
+namespace AutoBattle {
+
+#ifdef EP_DEBUG_AUTOBATTLE
+template <typename... Args>
+static void DebugLog(const char* fmt, Args&&... args) {
+	Output::Debug(fmt, std::forward<Args>(args)...);
+}
+#else
+template <typename... Args>
+static void DebugLog(const char*, Args&&...) {}
+#endif
+
+constexpr decltype(RpgRtCompat::name) RpgRtCompat::name;
+constexpr decltype(AttackOnly::name) AttackOnly::name;
+constexpr decltype(RpgRtImproved::name) RpgRtImproved::name;
+
+std::unique_ptr<AlgorithmBase> CreateAlgorithm(StringView name) {
+	if (Utils::StrICmp(name, RpgRtImproved::name) == 0) {
+		return std::make_unique<RpgRtImproved>();
+	}
+	if (Utils::StrICmp(name, AttackOnly::name) == 0) {
+		return std::make_unique<AttackOnly>();
+	}
+	if (Utils::StrICmp(name, RpgRtCompat::name) != 0) {
+		static bool warned = false;
+		if (!warned) {
+			Output::Debug("Invalid AutoBattle algo name `{}' falling back to {} ...", name, RpgRtCompat::name);
+			warned = true;
+		}
+	}
+	return std::make_unique<RpgRtCompat>();
+}
+
+void AlgorithmBase::SetAutoBattleAction(Game_Actor& source) {
+	vSetAutoBattleAction(source);
+	if (source.GetBattleAlgorithm() == nullptr) {
+		source.SetBattleAlgorithm(std::make_shared<Game_BattleAlgorithm::NoMove>(&source));
+	}
+}
+
+void RpgRtCompat::vSetAutoBattleAction(Game_Actor& source) {
+	SelectAutoBattleActionRpgRtCompat(source, Game_Battle::GetBattleCondition());
+}
+
+void AttackOnly::vSetAutoBattleAction(Game_Actor& source) {
+	SelectAutoBattleAction(source, Game_Battler::WeaponAll, Game_Battle::GetBattleCondition(), false, false, false, false);
+}
+
+void RpgRtImproved::vSetAutoBattleAction(Game_Actor& source) {
+	SelectAutoBattleAction(source, Game_Battler::WeaponAll, Game_Battle::GetBattleCondition(), true, false, false, false);
+}
+
+double CalcSkillHealAutoBattleTargetRank(const Game_Actor& source, const Game_Battler& target, const lcf::rpg::Skill& skill, bool apply_variance, bool emulate_bugs) {
+	assert(skill.type == lcf::rpg::Skill::Type_normal || skill.type >= lcf::rpg::Skill::Type_subskill);
+	assert(skill.scope == lcf::rpg::Skill::Scope_self || skill.scope == lcf::rpg::Skill::Scope_ally || skill.scope == lcf::rpg::Skill::Scope_party);
+
+	const double src_max_sp = source.GetMaxSp();
+	const double tgt_max_hp = target.GetMaxHp();
+	const double tgt_hp = target.GetHp();
+
+	if (target.GetHp() > 0) {
+		// Can the skill can heal the target?
+		if (!skill.affect_hp) {
+			return 0.0;
+		}
+
+		const double base_effect = Algo::CalcSkillEffect(source, target, skill, apply_variance);
+		const double max_effect = std::min(base_effect, tgt_max_hp - tgt_hp);
+
+		auto rank = static_cast<double>(max_effect) / static_cast<double>(tgt_max_hp);
+		if (src_max_sp > 0) {
+			const double cost = source.CalculateSkillCost(skill.ID);
```

Comment:
For ``CalculateSkillCost`` RPG_RT appears to ignore the "Half SP" (or I missed it). Maybe also emulate this bug?

Reply:
Nice catch, you're right this is another RPG_RT bug